### PR TITLE
[WIP] CMake + Github Actions

### PR DESCRIPTION
**Pull request by helloer**
This PR should significantly improve the build process by automating a lot of build steps with CMake.

Key points:
- `cmake && make` compiles client and server
- On the first run it will download base repository, create soldat.smod archive and copy other necessary files into build dir
- `FreeType` `OpenAL` `FreeType` have been added as required packages
- `stb` and `GameNetworkingSockets` will be built automatically and linked to the project

Build steps for Ubuntu(Linux):
- `sudo apt-get install build-essential g++ cmake git fpc libprotobuf-dev protobuf-compiler libssl-dev libsdl2-dev libopenal-dev libphysfs-dev libfreetype6`
- `mkdir build && cd build`
- `cmake ..`
- `make`

Build steps for macOS:
- `brew install openssl@1.1 protobuf fpc cmake sdl2 physfs freetype2`
- `mkdir build && cd build`
- `export PKG_CONFIG_PATH=$PKG_CONFIG_PATH:/usr/local/opt/openssl@1.1/lib/pkgconfig`
- `cmake -DOPENSSL_ROOT_DIR=$(brew --prefix openssl@1.1) ..`
- `make`

Build steps for Windows:
- Install [freepascal](https://sourceforge.net/projects/freepascal/files/Win32/3.0.4/) (Install first `fpc-3.0.4.i386-win32.exe`  and then `fpc-3.0.4.i386-win32.cross.x86_64-win64.exe`)
- Install [Visual Studio with C++ compiler/build tools](https://visualstudio.microsoft.com/en) and [vcpkg](https://github.com/Microsoft/vcpkg)
- open `Developer command prompt for visual studio`
-  `vcpkg.exe --triplet x64-windows install sdl2 physfs openssl protobuf freetype openal-soft`
- `set PATH=%PATH%;C:\fpc\3.0.4\bin\i386-win32`
- `set OPENSSL_ROOT_DIR=C:\vcpkg\installed\x64-windows`
- `set PHYSFSDIR=C:\vcpkg\installed\x64-windows`
- `mkdir build`
- `cd build`
- `cmake -G "NMake Makefiles" -DCROSS_WINDOWS_64=1 -DCMAKE_TOOLCHAIN_FILE="C:\vcpkg\scripts\buildsystems\vcpkg.cmake"  -DSDL2_BUILDING_LIBRARY=1 ..`
- `nmake`

Available CMake flags:

```
Flag name - Type - Default value - Description
BUILD_CLIENT:BOOL=ON - Build client
BUILD_SERVER:BOOL=ON - Build server
BUILD_SCRIPTCORE:BOOL=ON - Build with scripting engine
BUILD_RCON:BOOL=ON - Build with RCON
BUILD_SMOD:BOOL=ON - Fetches base repository and creates soldat.smod
BUILD_STEAM:BOOL=OFF - Steam version
```

Cross-compilation:
```
CROSS_DARWIN_32:BOOL=OFF - Sets target OS to macOS 32 bit
CROSS_DARWIN_64:BOOL=OFF - Sets target OS to macOS 64 bit
CROSS_LINUX_32:BOOL=OFF -  Sets target OS to linux 32 bit
CROSS_LINUX_64:BOOL=OFF -  Sets target OS to linux 64 bit
CROSS_WINDOWS_32:BOOL=OFF -  Sets target OS to windows 32 bit
CROSS_WINDOWS_64:BOOL=OFF -  Sets target OS to windows 64 bit
```

On top of that, this PR also adds basic CI with Github Actions to verify if cmake works properly. You can check https://github.com/helloer/soldat/actions to see how it works.

This work is based on [hedgewars](https://github.com/hedgewars/hw) and [this snippet](https://github.com/ValveSoftware/GameNetworkingSockets/issues/114#issuecomment-625211541)

### CODE REVIEW

**Comment by moep** [comment]
Here are the packages needed for compiling on Arch Linux:

```
pacman -Syy
pacman -S base-devel cmake fpc protobuf openssl sdl2 openal physfs freetype2
```

**Comment by zenyd** [comment]
I'm trying to compile on windows but on step `cmake -G "NMake Makefiles" -DCROSS_WINDOWS_64=1 -DCMAKE_TOOLCHAIN_FILE="C:\vcpkg\scripts\buildsystems\vcpkg.cmake" -DSDL2_BUILDING_LIBRARY=1 ..`

I get the error `CMake Error at shared/libs/GameNetworkingSockets/build/gns_git/CMakeLists.txt:100 (message):
  Cannot find EVP_MD_CTX_free in OpenSSL headers/libs for the target
  architecture.  Check that you're using OpenSSL 1.1.0 or later.`

Do you know how to resolve this?

**Comment by helloer** [comment]
Did you install dependencies with vcpkg (`vcpkg.exe --triplet x64-windows install sdl2 physfs openssl protobuf freetype openal-soft`)? Did you set OPENSSL_ROOT_DIR(make sure path is valid)?

**Comment by zenyd** [comment]
Yes I did, albeit my path for vcpkg differs (it's in my documents).

Here's the [CMakeError.log](https://github.com/Soldat/soldat/files/4729621/CMakeError.log). It's in german, but I can help translating if needed.

**Comment by helloer** [comment]
`C:\PROGRA~2\MICROS~1\2019\COMMUN~1\VC\Tools\MSVC\1426~1.288\bin\Hostx86\x86\cl.exe @C:\Users\zenyd\AppData\Local\Temp\nm9375.tmp`
` warning LNK4272:Bibliothekcomputertyp "x64" steht in Konflikt mit dem Zielcomputertyp "x86"`
This looks like cmake is using a 32-bit MSVC, try to run [64bit version of VS command prompt](https://docs.microsoft.com/sk-sk/cpp/build/how-to-enable-a-64-bit-visual-cpp-toolset-on-the-command-line?view=vs-2015) or if your system is 32-bit then you have to use 32-bit dependencies (`vcpkg.exe --triplet x86-windows install sdl2 physfs openssl protobuf freetype openal-soft`)

**Comment by zenyd** [comment]
So I copyed vcpkg to C:\ but it still wouldn't work. After I installed the x86 packages using `vcpkg.exe install sdl2 physfs openssl protobuf freetype openal-soft` I was able to run cmake. It's curious the x64 build would require x86 packages.

But on running nmake I ran into another problem when linking `[ 87%] Linking CXX shared library ..\..\..\..\..\..\bin\GameNetworkingSockets.dll`. ([build_errors.txt](https://github.com/Soldat/soldat/files/4731373/build_errors.txt)). It's complaining it can't resolve external symbols probably relating to protobuf. I tried both 32 and 64 bit builds.

**Comment by helloer** [comment]
Looks like paths for protobuf libs are hardcoded for x64 version in [shared/libs/GameNetworkingSockets/CMakeLists.txt](https://github.com/Soldat/soldat/blob/20d9ce928824c8db420a1bb53b3827a7c2956f27/shared/libs/GameNetworkingSockets/CMakeLists.txt#L32-L39), you need to change `x64-windows` to `x86-windows`. I'll fix it soon on my branch.

**Comment by zenyd** [comment]
yeah that helped, thanks! Does this mean when compiling in 64-bit this only applies to client and server and other libraries are linked using 32-bit versions?

**Comment by rikimbo** [comment]
I ran into a linker error when following these instructions on Ubuntu 20.04. Looks like the error stems from a (very) recent change in GameNetworkingSockets master that changed the callback mechanisms. Among other things, one of their API calls was incremented from v008 to v009: https://github.com/ValveSoftware/GameNetworkingSockets/commit/4c71718263b184309e29318005f1a1e5a4b94959#diff-d0302399d0370b4f3f9924df410d55c3R22

I was able to get this building successfully again by checking out an earlier commit of GameNetworkingSockets.

**Comment by Mailaender** [comment]
I can confirm that this successfully builds against GameNetworkingSockets 1.2.0 on @openSUSE so this already resolves https://github.com/Soldat/soldat/issues/33#issuecomment-752973800 for me. https://build.opensuse.org/request/show/859618 However it immediately crashes when trying to start the game:

```
./soldat_x64  
Cannot connect to server socket err
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
An unhandled exception occurred at $00007F0C1EFAA8B8:
EAccessViolation: 
  $00007F0C1EFAA8B8
```

I don't know if it is related.

File: cmake_modules/utils.cmake [hunk 1]
```diff
@@ -0,0 +1,57 @@
+macro(smod_builder)
+  include(ExternalProject)
+
+  ExternalProject_Add(
+    soldatbase
+    PREFIX "dist"
+    GIT_REPOSITORY https://github.com/Soldat/base.git
```
Comment (Mailaender) [inline cmake_modules/utils.cmake]: Doing a `git clone` during build is not something you want on any serious Linux distribution. Submodule it?

**Comment by rzaba0** [comment]
> 
> 
> I can confirm that this successfully builds against GameNetworkingSockets 1.2.0 on @openSUSE so this already resolves [#33 (comment)](https://github.com/Soldat/soldat/issues/33#issuecomment-752973800) for me. https://build.opensuse.org/request/show/859618 However it immediately crashes when trying to start the game:
> 
> ```
> ./soldat_x64  
> Cannot connect to server socket err
> Cannot connect to server request channel
> jack server is not running or cannot be started
> JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
> JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
> An unhandled exception occurred at $00007F0C1EFAA8B8:
> EAccessViolation: 
>   $00007F0C1EFAA8B8
> ```
> 
> I don't know if it is related.

I think you have to start the `./soldatserver` first, and then join it with `./soldat_x64 -join 127.0.0.1 23073`

**Comment by Mailaender** [comment]
I see. That also crashes as it does not adhere to filesystem hierarchy standards and tries to write where it can't

```
Unable to set fs_userpath: 
No such config file: server.cfg
```

```
An unhandled exception occurred at $0000000000435D25:
Exception: Failed to create configs/banned.txt
  $0000000000435D25
  $000000000042A46B
  $000000000040309E

An unhandled exception occurred at $000000000041B997:
EAccessViolation: Access violation
  $000000000041B997
  $0000000000418A11
  $000000000040309E
```

File: cmake_modules/utils.cmake (hunk 1 above)
Comment (BranDougherty) [reply cmake_modules/utils.cmake]: Resolved by afa6826472d88b26861b82687abf73d2b6a5cee5. We still do a "shallow" git clone for GameNetworkingSockets but I've been having trouble with it and might replace it with a download of the release anyways.

**Comment by BranDougherty** [comment]
As far as I can tell all issues brought up which are within the scope of this PR have been resolved. I'm sure we will find and fix more problems with this code, but it's been open for over a year now, while being a clear improvement over the makefiles which some people have struggled with. So in order to not let the perfect be the enemy of the good I'm merging this. @helloer if you can check the "Allow edits" button for this PR I can do the last thing needed to merge this (update the soldat asset download URL and soldat.smod checksum to match the official base repo's release).

**Review by BranDougherty — APPROVED** [review]

**Pull request merged**